source/extensions/quic_listeners/quiche/envoy_quic_utils.cc

```diff
@@ -124,5 +124,67 @@ createConnectionSocket(Network::Address::InstanceConstSharedPtr& peer_addr,
   return connection_socket;
 }
 
+bssl::UniquePtr<X509> parseDERCertificate(const std::string& der_bytes,
+                                          std::string* error_details) {
+  const uint8_t* data;
+  const uint8_t* orig_data;
+  orig_data = data = reinterpret_cast<const uint8_t*>(der_bytes.data());
+  bssl::UniquePtr<X509> cert(d2i_X509(nullptr, &data, der_bytes.size()));
+  if (!cert.get()) {
+    *error_details = "d2i_X509: fail to parse DER";
+    return nullptr;
+  }
+  if (data < orig_data || static_cast<size_t>(data - orig_data) != der_bytes.size()) {
+    *error_details = "There is trailing garbage in DER.";
+    return nullptr;
+  }
+  return cert;
+}
+
+int deduceSignatureAlgorithmFromPublicKey(const EVP_PKEY* public_key, std::string* error_details) {
+  int sign_alg = 0;
+  const int pkey_id = EVP_PKEY_id(public_key);
+  switch (pkey_id) {
+  case EVP_PKEY_EC: {
```

Comment:
Add test coverage for this case, including the error code for wrong curve/group

Reply:
done